### PR TITLE
DPDK_Article: for Sebastian review

### PR DESCRIPTION
**Pull request by tmulquee**

### CODE REVIEW

File: Validating DPDK Performance on OpenShift.md [hunk 1]
```diff
@@ -0,0 +1,652 @@
+# Validating DPDK performance on OpenShift
+This article describes the build and deployment of a traffic generating application inside a container. The traffic generator is used with:
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: ...traffic-generating application? (Not certain)
Is it OK to have build as a verb here? Maybe creation or generation of...?
We shouldn't have truncated introductory sentences according to style guide. So it should be something like: The traffic generator is used with the following infrastructure (or functionality or elements or...)

File: Validating DPDK Performance on OpenShift.md [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+- Hardware networking
+- Cloud elements
+- Physical and virtual functions
+- Test application
+
+to validate Data Plane Development Kit (DPDK) line rate performance on OpenShift Container Platform.
+## Traffic testing environment
+The following diagram shows the components of a traffic testing environment.
+![Traffic testing environment](traffic_testing.png "Traffic testing environment")
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md:10]: Some of these list item descriptions start with upper-case after the colon, and others, lower-case. You need to be consistent.

File: Validating DPDK Performance on OpenShift.md [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+- **Traffic generator**: an application that can generate high-volume packet traffic.
+- **SR-IOV-supporting NIC**: A network interface card compatible with Single Root I/O Virtualization. The card runs virtual functions on a physical interface.
+- **PF**: Physical Function: a PCI Express (PCIe) function of a network adapter that supports the single root I/O virtualization (SR-IOV) interface.
+- **VF**: Virtual Function (VF): a lightweight PCIe function on a network adapter that supports Single Root I/O virtualization (SR-IOV). The VF is associated with the PCIe Physical Function (PF) on the network adapter, and represents a virtualized instance of the network adapter.
+- **Switch**: Network switch. Nodes can also be connected back-to-back.
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md:15]: A/the network switch (for consistency). Relationship between  the two sentences 'Network switch.' and 'Nodes can also be...' isn't clear.

File: Validating DPDK Performance on OpenShift.md [hunk 4]
```diff
@@ -0,0 +1,652 @@
...
+- **`testpmd`**: An example application included with DPDK. The `testpmd` application can be used to test the DPDK in a packet forwarding mode. `testpmd` is also an example of how to build a fully-fledged application using the DPDK SDK.
+- **worker 0** and **worker 1**: OpenShift Container Platform nodes.
+
+## Overview
+For this stage of the development of the test environment, you must:
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: Change to something like: For this stage of the development of the test environment, you must do the following:
According to the Style Guide, we must have complete introductory sentences.

File: Validating DPDK Performance on OpenShift.md [hunk 5]
```diff
@@ -0,0 +1,652 @@
...
+do
+  CPU="$CPU,$i"
+done
+CPU="${CPU:1}"
+echo $CPU
+echo "Socket:" ${SOCKET}
+envsubst <  /opt/templates/trex_cfg.yaml > "/etc/trex_cfg.yaml"
+```
+## Deploy an OpenShift Container Platform cluster
+You must deploy and configure both the Node Tuning Operator and the SR-IOV Network Operator. See the [OpenShift Container Platform 4.10 Documentation](https://docs.openshift.com/container-platform/4.10/welcome/index.html) for more information. See also the Additional Resources at the end of this article.
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: See also Additional Resources .... (without 'the')

File: Validating DPDK Performance on OpenShift.md [hunk 6]
```diff
@@ -0,0 +1,652 @@
...
+## Commands for TRex
+`Exect` into the trex pod, and start the connection to TRex. You can then load the traffic generator.
+```sh
+./trex-console
+> tui
+> # start -f /opt/tests/testpmd.py -m <number-of-packets> -p <ports to use>
+> start -f /opt/tests/testpmd.py -m 24mpps -p 0
+> stop -a
+```
+This code example runs the traffic generator sending 24 million packets per second using port 0 and receiving from port 1.
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: Comma after generator?

File: Validating DPDK Performance on OpenShift.md [hunk 7]
```diff
@@ -0,0 +1,652 @@
...
+if [[ ${NODE} -eq 1 ]]
+then
+export SOCKET="1"
+fi
+
+echo "Socket:" ${SOCKET}
+envsubst <  /opt/templates/trex_cfg.yaml > "/etc/trex_cfg.yaml"
+```
+## Test configuration
+The test is injected to the container using a `configmap`. Using this method, you do not need to recreate the pod if you change the code. Apply the `configmap` again and `k8s` will mount the new files.
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: injected to or injected into?

File: Validating DPDK Performance on OpenShift.md [hunk 8]
```diff
@@ -0,0 +1,652 @@
...
+ name: hugepages
+```
+**Description**
+- `PCIDEVICE_OPENSHIFT_IO_DPDK_NIC_1`: The PCI deviceID for the first network.
+- `PCIDEVICE_OPENSHIFT_IO_DPDK_NIC_2`: The PCI deviceID for the second network.
+- `Trex-config-template`: the template configuration. Create this before starting the `trex` process.
+
+## TRex config script
+
+Use the following script to configure and launch TRex.
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: Sometimes you have a colon after the introductory sentence, sometimes a full stop, and sometimes nothing. Be consistent.

File: Validating DPDK Performance on OpenShift.md [hunk 9]
```diff
@@ -0,0 +1,652 @@
...
+ configMap:
+   name: trex-tests
+   - emptyDir:
+   medium: HugePages
+ name: hugepages
+```
+**Description**
+- `PCIDEVICE_OPENSHIFT_IO_DPDK_NIC_1`: The PCI deviceID for the first network.
+- `PCIDEVICE_OPENSHIFT_IO_DPDK_NIC_2`: The PCI deviceID for the second network.
+- `Trex-config-template`: the template configuration. Create this before starting the `trex` process.
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: The template configuration (capital T). Be consistent across the lists with this.

File: Validating DPDK Performance on OpenShift.md [hunk 10]
```diff
@@ -0,0 +1,652 @@
...
+   defaultHugepagesSize: 1G
+   pages:
+   - count: 32
+ size: 1G
+ numa:
+   topologyPolicy: "single-numa-node"
+ nodeSelector:
+   node-role.kubernetes.io/worker-cnf: ""
+```
+**Description**
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md:98]: The list below is not parallel - have forwarded you a mail about this.

File: Validating DPDK Performance on OpenShift.md [hunk 11]
```diff
@@ -0,0 +1,652 @@
+# Validating DPDK performance on OpenShift
+This article describes the build and deployment of a traffic generating application inside a container. The traffic generator is used with:
+- Hardware networking
+- Cloud elements
+- Physical and virtual functions
+- Test application
+
+to validate Data Plane Development Kit (DPDK) line rate performance on OpenShift Container Platform.
+## Traffic testing environment
+The following diagram shows the components of a traffic testing environment.
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: traffic-testing environment perhaps (not certain)?

File: Validating DPDK Performance on OpenShift.md [hunk 12]
```diff
@@ -0,0 +1,652 @@
...
+# wild second XL710 mac
+   mac_telco0 = '60:00:00:00:00:01'
+   # we don't care of the IP in this phase
+   ip_telco0  = '10.0.0.1'
+   # wild first XL710 mac
+   mac_telco1 = '60:00:00:00:00:02'
+   ip_telco1 = '10.1.1.1'
+```
+## Commands for TRex
+`Exect` into the trex pod, and start the connection to TRex. You can then load the traffic generator.
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: Everywhere else you wrote TRex pod.

File: Validating DPDK Performance on OpenShift.md (hunk 6 above)
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: Comma after generator?

File: Validating DPDK Performance on OpenShift.md (hunk 1 above)
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: Perhaps: The traffic generator validates Data Plane Development Kit (DPDK) line rate performance on OpenShift Container Platform. It is used with the following network elements:

File: Validating DPDK Performance on OpenShift.md [hunk 13]
```diff
@@ -0,0 +1,652 @@
+# Validating DPDK performance on OpenShift
+This article describes the build and deployment of a traffic generating application inside a container. The traffic generator is used with:
+- Hardware networking
+- Cloud elements
+- Physical and virtual functions
+- Test application
+
+to validate Data Plane Development Kit (DPDK) line rate performance on OpenShift Container Platform.
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: Move up before list.

File: Validating DPDK Performance on OpenShift.md [hunk 14]
```diff
@@ -0,0 +1,652 @@
...
+- **worker 0** and **worker 1**: OpenShift Container Platform nodes.
+
+## Overview
+For this stage of the development of the test environment, you must:
+1. Build the TRex container image.
+2. Run the TRex config script.
+3. Deploy an OpenShift Container Platform cluster.
+4. Define a TRex pod.
+
+## Build the TRex container image
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: Shouldn't headings be gerunds? Building the TRex container image, Running the TRex Config script, Deploying an OpenShift container platform, etc.

File: Validating DPDK Performance on OpenShift.md [hunk 15]
```diff
@@ -0,0 +1,652 @@
...
+- Physical and virtual functions
+- Test application
+
+to validate Data Plane Development Kit (DPDK) line rate performance on OpenShift Container Platform.
+## Traffic testing environment
+The following diagram shows the components of a traffic testing environment.
+![Traffic testing environment](traffic_testing.png "Traffic testing environment")
+- **Traffic generator**: an application that can generate high-volume packet traffic.
+- **SR-IOV-supporting NIC**: A network interface card compatible with Single Root I/O Virtualization. The card runs virtual functions on a physical interface.
+- **PF**: Physical Function: a PCI Express (PCIe) function of a network adapter that supports the single root I/O virtualization (SR-IOV) interface.
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: Physical Function (PF): A PCI Express...

File: Validating DPDK Performance on OpenShift.md [hunk 16]
```diff
@@ -0,0 +1,652 @@
...
+- **VF**: Virtual Function (VF): a lightweight PCIe function on a network adapter that supports Single Root I/O virtualization (SR-IOV). The VF is associated with the PCIe Physical Function (PF) on the network adapter, and represents a virtualized instance of the network adapter.
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: Remove the first VF and bold Virtual Function (VF)

File: Validating DPDK Performance on OpenShift.md [hunk 17]
```diff
@@ -0,0 +1,652 @@
...
+- **Switch**: Network switch. Nodes can also be connected back-to-back.
+- **`testpmd`**: An example application included with DPDK. The `testpmd` application can be used to test the DPDK in a packet forwarding mode. `testpmd` is also an example of how to build a fully-fledged application using the DPDK SDK.
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md:16]: Why is this a different font?

File: Validating DPDK Performance on OpenShift.md [hunk 18]
```diff
@@ -0,0 +1,652 @@
...
+- **worker 0** and **worker 1**: OpenShift Container Platform nodes.
+
+## Overview
```
Comment (tshwartz) [inline Validating DPDK Performance on OpenShift.md]: Perhaps this should be Procedure or Running the validation. Overview in the middle looks a bit weird.

File: Validating DPDK Performance on OpenShift.md (hunk 17 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md:16]: It's a command line utility, we use monospace for these.

File: Validating DPDK Performance on OpenShift.md (hunk 1 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md]: Added "the" in front of deployment so they are both nouns.

Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md]: I like it. Two short sentences are almost always better than one long one.

File: Validating DPDK Performance on OpenShift.md (hunk 11 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md]: Hyphen works better, agreed.

File: Validating DPDK Performance on OpenShift.md (hunk 2 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md:10]: Done.

File: Validating DPDK Performance on OpenShift.md (hunk 15 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md]: Done.

File: Validating DPDK Performance on OpenShift.md (hunk 16 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md]: Done.

File: Validating DPDK Performance on OpenShift.md (hunk 4 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md]: Done.

File: Validating DPDK Performance on OpenShift.md (hunk 14 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md]: Yes. Changed all headings to gerunds.

File: Validating DPDK Performance on OpenShift.md (hunk 5 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md]: Done.

File: Validating DPDK Performance on OpenShift.md (hunk 10 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md:98]: Thanks. Applied parallelism.

File: Validating DPDK Performance on OpenShift.md (hunk 9 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md]: Done.

File: Validating DPDK Performance on OpenShift.md (hunk 8 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md]: Standardized on colons.

File: Validating DPDK Performance on OpenShift.md (hunk 7 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md]: "injected to" and "injected into" both occur in the literature of Injection Containers. "injected into" makes more sense imho.

File: Validating DPDK Performance on OpenShift.md (hunk 12 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md]: Fixed.

File: Validating DPDK Performance on OpenShift.md (hunk 6 above)
Comment (tmulquee) [reply Validating DPDK Performance on OpenShift.md]: Done.

**Review by SchSeba — APPROVED** [review]

**Pull request merged**